api/src/main/java/org/apache/iceberg/catalog/TransactionalCatalog.java

```diff
@@ -0,0 +1,122 @@
+ *     determines a deadlock between two transactions has occurred).
+ * <li>The underlying implementation determines that the transaction can no longer complete
+ *     successfully.
+ * </ul>
+ *
+ * <p>When one of the items above occurs, the transaction is no longer valid. Further use
+ * of the transaction will result in a {@link IllegalStateException} being thrown.
+ *
+ * <p>Nested transactions such as creating a new table may fail. Those failures alone do
+ * not necessarily result in a failure of the catalog-level transaction.
+ *
+ * <p>Implementations of {@code TransactionalCatalog} are responsible for monitoring all
+ * table level operations that are spawned from this catalog and ensure that all nested
+ * transactions that are completed successfully are either exposed atomically or not.
+ *
+ */
+public interface TransactionalCatalog extends Catalog, AutoCloseable {
+
+  /**
+   * An internal identifier associated with this transaction.
+   * @return An internal identifier.
+   */
+  String transactionId();
+
+  /**
+   * Return the current {@code IsolationLevel} for this transaction.
+   * @return The IsolationLevel for this transaction.
+   */
+  IsolationLevel isolationLevel();
+
+  /**
+   * Return the {@link LockingMode} for this transaction.
+   * @return The LockingMode for this transaction.
+   */
+  LockingMode lockingMode();
+
+  /**
+   * Whether the current transaction is still active/open.
+   * @return True until a terminal action occurs.
+   */
+  boolean active();
+
+  /**
+   * Aborts the set of operations here and makes this TransactionalCatalog inoperable.
+   *
+   * <p>Once called, no further operations can be done against this catalog. If any
+   * operations are attempted, {@link IllegalStateException} will be thrown.
+   */
+  void rollback();
+
+  /**
+   * Commit the pending changes from all nested transactions against the Catalog.
+   *
+   * <p>Once called, no further operations can be done against this catalog. If any
+   * operations are attempted, {@link IllegalStateException} will be thrown.
+   *
+   * @throws CommitFailedException If the updates cannot be committed due to conflicts.
+   */
+  void commit();
+
+  /**
+   * Close out all resources associated with a transaction.
+   *
+   * <p>This will do a conditional rollback if neither {@code commit} nor {@code rollback}
+   * were called. Standard usage looks like:
+   * <pre>
+   * try(TransactionalCatalog tx = catalog.createTransaction(IsolationLevel.READ_COMMITTED)) {
+   *  doOp1(tx);
+   *  doOp2(tx);
+   *  tx.commit();
+   * }
+   * </pre>
+   * This pattern is designed such that if {@code doOp1()} or {@code doOp2()} throw an exception,
+   * the transaction will be automatically rolled back. If both operations complete successfully,
+   * the close will only close any remaining open resources associated with the transaction.
+   */
+  @Override
+  default void close() {
+    commit();
+    close();
```

Comment:
Wouldn't this cause stackoverflow?

Reply:
That's funny. Yes, I was checking the javadoc formatting in the ide and forgot to remove the extraneous close.